src/datastructures.ts

```diff
@@ -1,25 +1,40 @@
-export class Cons<A> {
-  constructor(
-    public readonly value: A,
-    public readonly tail: Cons<A>,
-    public readonly isNil: boolean
-  ) {}
-  *[Symbol.iterator](): IterableIterator<A> {
-    let head: Cons<A> = this;
-    while (head.isNil === false) {
-      const v = head.value;
-      head = head.tail;
-      yield v;
-    }
-  }
+export interface ConsNil {
+  readonly isNil: true;
+  [Symbol.iterator](): Generator<never, void, undefined>;
 }
 
-export const nil: Cons<any> = new Cons(undefined, undefined, true);
+export interface ConsValue<A> {
+  readonly isNil: false;
+  readonly value: A;
+  readonly tail: Cons<A>;
+  [Symbol.iterator](): Generator<A, void, undefined>;
+}
+
+export type Cons<A> = ConsNil | ConsValue<A>;
 
-export function cons<A>(value: A, tail: Cons<A> = nil): Cons<A> {
-  return new Cons(value, tail, false);
+function* generator<A>(this: Cons<A>) {
+  let head: Cons<A> = this;
+  while (!head.isNil) {
+    const v = head.value;
+    head = head.tail;
+    yield v;
+  }
 }
 
+export const nil: Cons<never> = {
+  isNil: true,
+  [Symbol.iterator]: generator
+};
+
+export const cons = <A>(value: A, tail: Cons<A> = nil): Cons<A> => {
+  return {
+    isNil: false,
+    value,
+    tail,
+    [Symbol.iterator]: generator
+  };
+};
+
```

Comment:
Instead of this change, what about implementing the linked list as an abstract class called `List` which `Cons` and `Nil` extends? Then the iterator and the `isNil` property could sit on the abstract class. I also think that would fit better with the rest of the code where we use abstract classes and classes that extend them to represent something like algebraic data types/sum types.

Reply:
Thanks for the reply. Personally I just use tagged unions and simple constructors to encode ADT in TypeScript because it lets to check such unions exhaustively, sort of exhaustive pattern matching. But I'll check your suggestion, thanks!

Reply:
@paldepind Looks like it's impossible to declare Cons using abstract class, because it would require declaring `value` and `tail` fields on `Nil` with some "nully" values to make shared generator work. That would defeat the very purpose of this PR - get rid of such values.
Declaring Cons using tagged union OTOH allows to declare `value` and `tail` only when `isNil === true` (means using `isNil` as tag).